presto-hive/src/main/java/com/facebook/presto/hive/SortingFileWriter.java

```diff
@@ -0,0 +1,207 @@
+import io.airlift.units.DataSize;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.Path;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_TOO_MANY_BUCKET_SORT_FILES;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static com.facebook.presto.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static io.airlift.units.DataSize.Unit.MEGABYTE;
+import static java.util.Objects.requireNonNull;
+
+public class SortingFileWriter
+        implements HiveFileWriter
+{
+    private static final Logger log = Logger.get(SortingFileWriter.class);
+
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SortingFileWriter.class).instanceSize();
+
+    private final FileSystem fileSystem;
+    private final Path tempFilePrefix;
+    private final int maxTempFiles;
+    private final List<Type> types;
+    private final List<Integer> sortFields;
+    private final List<SortOrder> sortOrders;
+    private final HiveFileWriter outputWriter;
+    private final SortBuffer sortBuffer;
+    private final List<Path> tempFiles = new ArrayList<>();
+
+    public SortingFileWriter(
+            FileSystem fileSystem,
+            Path tempFilePrefix,
+            HiveFileWriter outputWriter,
+            DataSize maxMemory,
+            int maxTempFiles,
+            List<Type> types,
+            List<Integer> sortFields,
+            List<SortOrder> sortOrders,
+            PageSorter pageSorter)
+    {
+        checkArgument(maxTempFiles > 0, "maxTempFiles must be greater than zero");
+        this.fileSystem = requireNonNull(fileSystem, "fileSystem is null");
+        this.tempFilePrefix = requireNonNull(tempFilePrefix, "tempFilePrefix is null");
+        this.maxTempFiles = maxTempFiles;
+        this.types = ImmutableList.copyOf(requireNonNull(types, "types is null"));
+        this.sortFields = ImmutableList.copyOf(requireNonNull(sortFields, "sortFields is null"));
+        this.sortOrders = ImmutableList.copyOf(requireNonNull(sortOrders, "sortOrders is null"));
+        this.outputWriter = requireNonNull(outputWriter, "outputWriter is null");
+        this.sortBuffer = new SortBuffer(maxMemory, types, sortFields, sortOrders, pageSorter);
+    }
+
+    @Override
+    public long getWrittenBytes()
+    {
+        return outputWriter.getWrittenBytes();
+    }
+
+    @Override
+    public long getSystemMemoryUsage()
+    {
+        return INSTANCE_SIZE + sortBuffer.getRetainedBytes();
+    }
+
+    @Override
+    public void appendRows(Page page)
+    {
+        if (tempFiles.size() == maxTempFiles) {
```

Comment:
this seems like a strange place for this check

Reply:
I was trying to make this fail eagerly, but it actually doesn't work correctly anyway due to how the flush logic works.